src/Pim/Bundle/EnrichBundle/EventListener/Storage/RemoveCategoryFilterInJobInstanceSubscriber.php

```diff
@@ -0,0 +1,174 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Pim\Bundle\EnrichBundle\EventListener\Storage;
+
+use Akeneo\Component\Batch\Model\JobInstance;
+use Akeneo\Component\StorageUtils\Saver\BulkSaverInterface;
+use Akeneo\Component\StorageUtils\StorageEvents;
+use Doctrine\ORM\EntityRepository;
+use Pim\Component\Catalog\Model\CategoryInterface;
+use Symfony\Component\EventDispatcher\EventSubscriberInterface;
+use Symfony\Component\EventDispatcher\GenericEvent;
+
+/**
+ * @author    Nicolas Marniesse <nicolas.marniesse@akeneo.com>
+ * @copyright 2019 Akeneo SAS (http://www.akeneo.com)
+ * @license   http://opensource.org/licenses/osl-3.0.php  Open Software License (OSL 3.0)
+ */
+class RemoveCategoryFilterInJobInstanceSubscriber implements EventSubscriberInterface
+{
+    private const DEFAULT_CATEGORY_TREE_FILTER = ['master'];
+
+    /** @var EntityRepository */
+    private $repository;
+
+    /** @var BulkSaverInterface */
+    private $bulkSaver;
+
+    /** @var array */
+    private $computedCodes = [];
+
+    public function __construct(EntityRepository $repository, BulkSaverInterface $bulkSaver)
+    {
+        $this->repository = $repository;
+        $this->bulkSaver = $bulkSaver;
+    }
+
+    public static function getSubscribedEvents()
+    {
+        return [
+            StorageEvents::PRE_REMOVE      => 'computeAndHoldCategoryTreeCodes',
+            StorageEvents::PRE_REMOVE_ALL  => 'computeAndHoldCategoryTreeCodes',
+            StorageEvents::POST_REMOVE     => 'removeCategoryFilter',
+            StorageEvents::POST_REMOVE_ALL => 'removeCategoryFilters',
+        ];
+    }
+
+    /**
+     * We can not get the child's codes of a category after delete. So when a category is going to be deleted,
+     * we get all the child's codes linked to it and keep them in $computedCodes property.
+     *
+     * @param GenericEvent $event
+     * @return RemoveCategoryFilterInJobInstanceSubscriber
+     */
+    public function computeAndHoldCategoryTreeCodes(GenericEvent $event): RemoveCategoryFilterInJobInstanceSubscriber
+    {
+        $subject = $event->getSubject();
+        if ($subject instanceof CategoryInterface) {
```

Comment:
```suggestion
        if (!$subject instanceof CategoryInterface) {
            return;
        }
        $this->computedCodes[$subject->getCode()] = $this->getCodeAndChildrenCodes($subject);
```
I find it simpler to early retrun and not have to test for the interface in the rest of the code.

Reply:
I prefer early return too, but here we can't do that. The subject can be a category or an array of categories, the method handles both cases.

Reply:
The check will be more complicated as it apparently can be an array of CategoryInterface too, but I agree with you, early return are almost always easier to understand.